### PR TITLE
ID-158 Make Sam return NOT_FOUND if user not authorized to modify group

### PR DESCRIPTION
**Pull request by tlangs**
https://broadworkbench.atlassian.net/browse/ID-158

Sam should not squash unauthorized group modification into a `NO_CONTENT` response. This gives users and downstream tools the wrong idea of what's going on. We should instead return a `NOT_FOUND` response.

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://broadinstitute.github.io/dsp-appsec-security-risk-assessment/) and attached the result to the JIRA ticket

### CODE REVIEW

**Comment by andy7i** [comment]
drive by question because I was curious about this one. did you find anything relating to this theory?:
> guessing the 204s might be changes to make the ui work a certain way, and docs not updated (that’s one ui dev strategy to allow calls w/ no page nav)? so am wagering that docs are just out of date. don’t have full history without looking at commits, but we can do a documentation update with team when we have time to fill things in.

**Review by Shakespeared — APPROVED** [review]

**Comment by tlangs** [comment]
I'm not sure how this would be represented in the UI. The user wouldn't be able to even load a page for a resource they don't have access to, so there's no way they should be able to make this request from the UI.

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutesSpec.scala [hunk 1]
```diff
@@ -500,6 +500,15 @@ class ManagedGroupRoutesSpec extends AnyFlatSpec with Matchers with ScalatestRou
     }
   }
 
+  it should "respond with a 404 when the user does not have access to modify the group" in {
+    val samRoutes = TestSamRoutes(resourceTypes)
+    withUserNotInGroup(samRoutes){ nonMemberRoutes =>
+      Delete(s"/api/group/$groupId/admin/${samRoutes.user.email}") ~> nonMemberRoutes.route ~> check {
```
Comment (gpolumbo-broad) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutesSpec.scala:509]: You're making a `DELETE` request here, but the code you modified is for `PUT`.  This test might be worth keeping too, but make sure the `PUT` is also tested.

**Review by gpolumbo-broad — CHANGES_REQUESTED** [review]

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutesSpec.scala (hunk 1 above)
Comment (tlangs) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutesSpec.scala:509]: Ah shoot, you're totally right!

**Review by gpolumbo-broad — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala [hunk 2]
```diff
@@ -116,7 +116,8 @@ trait ManagedGroupRoutes extends SamUserDirectives with SecurityDirectives with
       samRequestContext: SamRequestContext): Route =
     requireAction(managedGroup, SamResourceActions.sharePolicy(accessPolicyName), samUser.id, samRequestContext) {
       withSubject(WorkbenchEmail(email), samRequestContext) { subject =>
-        complete(managedGroupService.addSubjectToPolicy(managedGroup.resourceId, accessPolicyName, subject, samRequestContext).map(_ => StatusCodes.NoContent))
+        complete(managedGroupService.addSubjectToPolicy(managedGroup.resourceId, accessPolicyName, subject, samRequestContext)
+          .map(if (_) StatusCodes.NoContent else StatusCodes.NotFound))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala:120]: I don't think this does what you want. The authorization check is the `requireAction` call above. `addSubjectToPolicy` returns false when no change was actually made (i.e. member was already in the groups), not when the caller lacks access.

**Pull request merged**